
# Base-N and Binary Number Systems

Numbers can be represented in a multiple different ways. In our everyday lives, we commonly use a base 10 number system in which coefficients ranging from 0 to 9 represents the digits of a power of ten. For example:

In [2]:
# Base 10
256 == 2*(10**2) + 5*(10**1) + 6*(10**0)

True

It is also possible for us to represent the same number in different numerical systems. For instance, here is 256 represented in base 2, also known as binary:

In [3]:
256 == 1*(2**8)

True

In python, there is an easy method for us to convert numbers to binary, simply by calling the bin(k) method, where k is some integer we wish to convert. An example of this is as follows:

In [4]:
bin(256)

'0b100000000'

Normal mathematical operations still hold in other numerical systems. To demonstrate, here is how we add would about adding two integers:

In [11]:
# Addition
x = 1*(2**8) + (1*(2**0) + 1*(2**1) + 1*(2**2) + 1*(2**3))

print(x)
256 + 15 == x

271


In [13]:
# Using python shortcut
a = bin(256)
b = bin(15)
c = a + b

print(c)

0b1000000000b1111


# Floating Point Numbers
Up until this point, we have only looked at Integers, but we have yet to talk about floating point numbers. Floating point numbers allow us to represent decimals by storing three key pieces of information:
- A sign indicator, s, indicates if the number is positive or negative
- A characteristic/exponent, e, represented as the power of 2
- A fraction, f, the coefficient of e

Python allocates 1 bit, or binary number, to the sign, 11 bits to the characteristic, and 54 bits to the fraction. The sum of these bits is 64, hence we label it 64 bit. We can check the floating point information in python using the following code:

In [15]:
import sys
sys.float_info

sys.float_info(max=1.7976931348623157e+308, max_exp=1024, max_10_exp=308, min=2.2250738585072014e-308, min_exp=-1021, min_10_exp=-307, dig=15, mant_dig=53, epsilon=2.220446049250313e-16, radix=2, rounds=1)

As we can see from the output, we are told that the maximum input value is roughly 1.7976931348623157e+308. This is derived from the maximum exponent being given as 2^(e) where e can be, at maximum, equal to 10. Thus, we find that 2^10 = 1024, and we find the maximum amount of data to be equal to 2^1024. We can see that python considers any number larger than this to be beyond its capacity, thus returns infinity. For example: 

In [18]:
x = 2e308
print(x)

inf


We can check for the spacing between large numbers uisng the following command:

In [19]:
import numpy as np

np.spacing(1e7)

1.862645149230957e-09

Checking to see if this is true is simple enough. If we add any number less than half of the value, we will simply have the original number:

In [20]:
1e7 == 1e7 + np.spacing(1e7)/10

True

# Rounding Errors
While working in python, it becomes clear quite quickly that we can run into issues when working with decimals. For instance:

In [23]:
x = 15.5
y = 3.255
z = x-y

12.245 == z

False

The code returned false due to the way that floating point values approximate numerical value. The work around for this is by using the round function, in this case rounding to three decimal places:

In [24]:
12.245 == round(z,3)

True

We can add and subtract decimal numbers from integers and see that python returns a floating point number:

In [25]:
15 + 1/3 - 1/3

15.0

However, we run into increasing problems as we increase the number of iterations. For instance:

In [27]:
total_i = 50000
x = 5

for i in range(total_i):
    x += 1/3
for i in range(total_i):
    x -= 1/3

print(x)
x == 5

4.999999999998293


False